### ISSUE TITLE
User acct deletion: endpoint that allows MDN users to delete acct/archive their contribs

### ISSUE THREAD

**Issue by tobinmori** [mdn/kuma#5862]
**Related stories:**
Parent: https://github.com/mdn/sprints/issues/1922
Parent: https://github.com/mdn/sprints/issues/1947
Frontend: https://app.zenhub.com/workspaces/mdn-dev-5d6ea9fb3fe11605a63cea86/issues/mdn/kuma/6071

**Acceptance:**
- [ ] Use case 1: Endpoint allows the ability for non-contributor users to delete their account from the Modal 
- [ ] Use case 2: Endpoint allows the ability for contributor users to delete their account from the Modal, BUT retain attributions on contributions with username intact
- [ ] Use case 3: Endpoint allows the ability for contributor users to delete their account from the Modal, BUT retain attributions as 'anonymous'

**Hypothetical Incoming Data Args**: 
- MDN uid, attributions_prefers_anon, action

**Comment by peterbe** [comment]
@atopal @tobinmori Who has the flow chart diagram we drew in the meeting room in Portland this summer? I'd love to get started on the back-end piece so I never have to manually deal with these requests any more.

**Comment by atopal** [comment]
I just checked, and I don't have a picture of that.

**Comment by tobinmori** [comment]
Just checked too @peterbe, unfortunately, I have 2 photos of boards that aren't this one.

**Comment by peterbe** [comment]
Note-to-self; here's the first question: https://mozilla.invisionapp.com/share/Y8SGMKQBCW3#/screens/370705982

**Comment by atopal** [comment]
If I remember correctly the main question was what happens when people don't want to be associated with their content anymore. Do we delete their account and assign the content to one anonymous account or whether we rename their account name to "anonymous" and leave it at that.

**Comment by tobinmori** [comment]
@escattone do you happen to the workflow image we took of this in Portland? 

If not, @peterbe will recreate the workflow and vet before implementing.

**Comment by escattone** [comment]
> If I remember correctly the main question was what happens when people don't want to be associated with their content anymore. Do we delete their account and assign the content to one anonymous account or do we rename their account name to "anonymous" and leave it at that?

Yes, this is what I remember as well, thanks @atopal. I remember @atopal saying that he didn't care as to which of the choices we made, but I also remember that we, the developers, felt that it might be best to use a single anonymous account?

**Comment by escattone** [comment]
> @escattone do you happen to the workflow image we took of this in Portland?

I thought I did, but unfortunately I don't. 😞

**Comment by atopal** [comment]
Confirming that I don't mind either way.

**Comment by tobinmori** [comment]
@atopal Re: archiving users, clarifying: are we assuming that this just happens automatically and is not a user choice?

**Comment by peterbe** [comment]
@atopal @escattone @tobinmori @schalkneethling Here's what I think we decided in Portland:
![Issue 5862(2)](https://user-images.githubusercontent.com/26739/69094079-b843c580-0a1d-11ea-8d6a-ec1e1fe923a6.png)


(Note-to-self; https://www.draw.io/#G1Jrf7H3YK5OtNz621dBff6M70vAzCUBTm)

3 things can happen when the user wants to delete their account:

1. They have 0 revision contributions so just wipe the user account off the face of the earth. 
2. They have >0 revision contributions, we anonymize the hell out of them. 
3. We donate all their revision contributions to a user called "anonymous", and wipe the user account off the face of the earth. 

An important implication is that, if I do this, the only thing that remains of users in option (2) is the
username. So, if I do this option, but then decide to change my mind again 6 months later, will it complain that username `peterbe` is taken and that I now need to pick `peterbe2` in my comeback? Or will it "revive" that scrubbed user account and its revisions?

Perhaps ^ question depends on trying out what actually happens and how `django-allauth` works. 

But generally, does this seem like a good plan?

**Comment by escattone** [comment]
@peterbe That seems like a good plan to me!

> So, if I do this option, but then decide to change my mind again 6 months later, will it complain that username peterbe is taken and that I now need to pick peterbe2 in my comeback? Or will it "revive" that scrubbed user account and its revisions?

If we tried to "revive" the account, how would we know if it's the same person? I suppose we'd have to save an email address (and so couldn't wipe everything except the username) and send a recovery code or something like that?

**Comment by schalkneethling** [comment]
> @peterbe That seems like a good plan to me!
> 
> > So, if I do this option, but then decide to change my mind again 6 months later, will it complain that username peterbe is taken and that I now need to pick peterbe2 in my comeback? Or will it "revive" that scrubbed user account and its revisions?
> 
> If we tried to "revive" the account, how would we know if it's the same person? I suppose we'd have to save an email address (and so couldn't wipe everything except the username) and send a recovery code or something like that?

Yup, we would have to keep the associated email address as well so we can do the whole account recovery flow.

**Comment by atopal** [comment]
We won't offer a way to get an old username, if you decided to delete your MDN account. If this turns out to be frequent and contentious, we can revise our implementation.

**Comment by atopal** [comment]
@peterbe yes, that diagram is what we had agreed on, IIRC. The only thing I can't remember is whether we agreed to change the username to "anonymous" or actually use the user anonymous to attribute those edits. I personally don't have an opinion either way.

**Comment by schalkneethling** [comment]
> @peterbe yes, that diagram is what we had agreed on, IIRC. The only thing I can't remember is whether we agreed to change the username to "anonymous" or actually use the user anonymous to attribute those edits. I personally don't have an opinion either way.

We decided to use the single, catch-all `anonymous` user for these instances.

**Comment by peterbe** [comment]
@tobinmori I think the sprint paperwork might need some re-wording. What I built in https://github.com/mdn/kuma/pull/6149 is entirely "self-sufficient" in the Wiki and it doesn't depend on any new UI. I.e. this isn't blocked in any way on that modal. 

Also, this Django functionality is tied to the regular Django patterns of the Wiki. 

I'm losing track of progress and stuff but some day, if the self-destruction is done from a SPA in React, it might need to an XHR request to something like `DELETE /api/v1/peterbe`. That day, we can tackle that as a separate issue/PR. 

Once this lands in `master` (and released consequently after) we'll cease forcing people to go to Bugzilla as they'll be entirely able to use wiki.developer.mozilla.org to delete their account.

**Comment by schalkneethling** [comment]
@peterbe I agree that the modal should not block this. It would be easy enough for me to rework the flow into a modal, and having this in production is beneficial to both end-users and staff. Thanks for getting this done 🎉
